mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala

```diff
@@ -18,118 +20,178 @@ final class PcSyntheticDecorationsProvider(
-      case TypeParameters(types, pos) if params.typeParameters() =>
-        val label = types.map(toLabel(_, pos)).mkString("[", ", ", "]")
-        val adjusted = pos.adjust(text)._1
-        decorations.add(
-          Decoration(
-            adjusted.focusEnd.toLsp,
-            label,
-            DecorationKind.TypeParameter
-          )
+      case ValueOf(label, pos) if params.implicitParameters() =>
+        inlayHints.add(
+          adjustPos(pos).focusEnd.toLsp,
+          LabelPart("(") :: LabelPart(label) :: List(LabelPart(")")),
+          InlayHintKind.Parameter
         )
-      case InferredType(tpe, pos) if params.inferredTypes() =>
-        val label = toLabel(tpe, pos)
-        val adjusted = pos.adjust(text)._1
-        if (decorations.containsDef(adjusted.end)) decorations
+      case TypeParameters(tpes, pos)
+          if params.typeParameters() && tpes.forall(_ != null) =>
+        val head :: rest =
+          tpes.map(tpe => toLabelParts(tpe.finalResultType, pos)): @unchecked
+        val label = (head :: rest.map(LabelPart(", ") :: _)).flatten
+        inlayHints.add(
+          adjustPos(pos).focusEnd.toLsp,
+          LabelPart("[") :: label ++ List(LabelPart("]")),
+          InlayHintKind.Type
+        )
+      case InferredType(tpe, pos) if params.inferredTypes() && tpe != null =>
+        val adjustedPos = adjustPos(pos).focusEnd
+        if (inlayHints.containsDef(adjustedPos.start)) inlayHints
         else
-          decorations.add(
-            Decoration(
-              adjusted.focusEnd.toLsp,
-              ": " + label,
-              DecorationKind.InferredType
-            ),
-            adjusted.end
-          )
-      case _ => decorations
+          inlayHints
+            .add(
+              adjustedPos.toLsp,
+              LabelPart(": ") :: toLabelParts(tpe.finalResultType, pos),
+              InlayHintKind.Type
+            )
+            .addDefinition(adjustedPos.start)
+      case _ => inlayHints
     }
 
   def traverse(
-      acc: Synthetics,
+      acc: InlayHints,
       tree: Tree
-  ): Synthetics = {
-    val decorations = collectDecorations(tree, acc)
-    tree.children.foldLeft(decorations)(traverse(_, _))
+  ): InlayHints = {
+    val inlayHints = collectDecorations(tree, acc)
+    tree.children.foldLeft(inlayHints)(traverse(_, _))
+  }
+
+  private def partsFromType(tpe: Type): List[LabelPart] = {
+    tpe
+      .collect {
+        case t: TypeRef if t.sym != NoSymbol =>
+          labelPart(t.sym)
+        case SingleType(_, sym) if sym != NoSymbol =>
+          labelPart(sym)
+      }
   }
 
+  private def toLabelParts(
+      tpe: Type,
+      pos: Position
+  ): List[LabelPart] = {
+    val context: Context = doLocateImportContext(pos)
+    val re: scala.collection.Map[Symbol, Name] = renamedSymbols(context)
+    val history = new ShortenedNames(
+      lookupSymbol = name =>
+        context.lookupSymbol(name, sym => !sym.isStale) :: Nil,
+      config = renameConfig,
```

Comment:
I think showing type as `ju.List` is better than `java.util.List`?

Reply:
It's just that this is an arbitrary rename made up by metals. It makes sense when you add an import but I'm not sure how intuitive it is when you just print the type. Just IMO, I'm fine with whatever you decide.

Reply:
If ju rename is present in the file then it would be good to use it otherwise I would show the full name. Co renameConfig is probably not too useful in this case